core/network/impl/stream_engine.hpp

```diff
@@ -0,0 +1,287 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_STREAM_ENGINE_HPP
+#define KAGOME_STREAM_ENGINE_HPP
+
+#include <unordered_map>
+
+#include "common/logger.hpp"
+#include "libp2p/connection/stream.hpp"
+#include "libp2p/host/host.hpp"
+#include "libp2p/peer/peer_info.hpp"
+#include "libp2p/peer/protocol.hpp"
+#include "network/helpers/scale_message_read_writer.hpp"
+#include "subscription/subscriber.hpp"
+#include "subscription/subscription_engine.hpp"
+
+namespace kagome::network {
+
+  struct StreamEngine final : std::enable_shared_from_this<StreamEngine> {
+    using PeerInfo = libp2p::peer::PeerInfo;
+    using PeerId = libp2p::peer::PeerId;
+    using Protocol = libp2p::peer::Protocol;
+    using Stream = libp2p::connection::Stream;
+    using Host = libp2p::Host;
+    using StreamEnginePtr = std::shared_ptr<StreamEngine>;
+
+    enum PeerType { kSyncing = 1, kReserved };
+
+   private:
+    using ProtocolMap = std::unordered_map<Protocol, std::shared_ptr<Stream>>;
+    using PeerMap = std::unordered_map<PeerInfo, ProtocolMap>;
+
+    struct ProtocolDescriptor {
+      std::reference_wrapper<ProtocolMap> proto_map;
+      PeerType type;
+    };
+
+   public:
+    StreamEngine(const StreamEngine &) = delete;
+    StreamEngine &operator=(const StreamEngine &) = delete;
+
+    StreamEngine(StreamEngine &&) = delete;
+    StreamEngine &operator=(StreamEngine &&) = delete;
+
+    ~StreamEngine() = default;
+    explicit StreamEngine(Host &host)
+        : host_{host}, logger_{common::createLogger("StreamEngine")} {}
+
+    template <typename... Args>
+    static StreamEnginePtr create(Args &&... args) {
+      return std::make_shared<StreamEngine>(std::forward<Args>(args)...);
+    }
+
+    void addReserved(const PeerInfo &peer_info,
+                     const Protocol &protocol,
+                     std::shared_ptr<Stream> stream) {
+      BOOST_ASSERT(!protocol.empty());
+
+      std::unique_lock cs(streams_cs_);
+      auto &protocols = reserved_streams_[peer_info];
+      BOOST_ASSERT(protocols.find(protocol) == protocols.end());
+      protocols.emplace(protocol, std::move(stream));
+    }
+
+    outcome::result<void> add(const Protocol &protocol,
+                              std::shared_ptr<Stream> stream) {
+      BOOST_ASSERT(stream);
+      BOOST_ASSERT(!protocol.empty());
+
+      OUTCOME_TRY(peer, from(stream));
+      bool existing = false;
+
+      std::unique_lock cs(streams_cs_);
+      forSubscriber(peer, protocol, [&](auto type, auto &subscriber) {
+        existing = true;
+        if (subscriber != stream) uploadStream(subscriber, std::move(stream));
```

Comment:
Not sure how forSubscriber() works exactly, but it seems suspicious that stream is moved here and also on line 84.